### DISCUSSION TITLE
gifs not displaying properly, gifDraw writer says example isn't respecting iX

### DISCUSSION THREAD

**Discussion by h3llb3nt** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#269] [Q&A] [answered]
im having some issues with the animatedGIFPanel example code. i've got one of the ESP32 Trinity boards hooked up to a 64x32 display and with the supplied example gifs things work great. however with my custom ones im getting odd display issues. i created a test 64x32 animated gif. here is the source file:
![test](https://user-images.githubusercontent.com/22822528/159335649-982ae92c-5860-40aa-95b5-8f6a76eed256.gif)
and here's what it looks like using the AnimatedGIFPanel.ino example code:
https://user-images.githubusercontent.com/22822528/159335774-9c367576-03c7-47a7-9724-9ac2d47aefeb.mov
i have others that don't work as well. this donkey kong gif for example.
![donkey_64](https://user-images.githubusercontent.com/22822528/159336487-9236519c-1a29-4ff4-8f0d-95468ff72600.gif)
which looks like this on the display:
https://user-images.githubusercontent.com/22822528/159337237-9665b871-fdcf-48d6-8546-a7adc9ff28b7.mov
i contacted Larry Bank the AnimatedGIF.h creator and he said "that code is ignoring the iX value in the GIFDraw callback. The iX/iY are the upper left corner of the delta frame (within the larger GIF canvas)." unfortunately im way out of my depth and was hoping someone could assist me in tweaking the code to work. 

i've also confirmed that its NOT my gifs themselves as they work fine using the same animatedGIF library on a pyportal. Larry mailed me back and said "The issue is that the iX/iY variables passed in the GIFDRAW callback refer to the upper left corner of the sub-image being drawn onto the canvas. Your code only uses the iY value, so the changes are all squished up against the left side by ignoring the iX value." however, i have NO idea what to adjust or how to.

**Comment by GrimorJander** [comment, marked as answer]
Seens stacked frames to me.
Try optimizing those gifs with Ezgif this way:  Ezgif/Frames/ mark "Don't stack frames" option and create new gif

**Reply by GrimorJander** [reply]
I just tried your Donkey Kong Gif and it does indeed have staked frames, displaying exactly the same as your example.
Then, I have allowed myself to de-stack and it already behaves as it should. Attached the corrected Gif.
![donkey](https://user-images.githubusercontent.com/27100419/194728094-48e5714a-e7c7-4c07-83de-169a30c9c89b.gif)
My advice with GifAnimation example is to put CLKPhase: True , remove the 8 seconds limit on the code and optimize properly all the images you send to device.
With SD card you will be able to play really bif gif animations (like 20Mb and 4 minutes animation).
For stacked images (that retain ghosting from previous frames) make this with EZgif:
Upload your image and go to Frames tab:
![1](https://user-images.githubusercontent.com/27100419/194728310-6e731717-69cc-4a8b-969c-7499bc277752.png)
then under efects on the botton of the page, mark "dont stack frames" Option and remake the gif.
![2](https://user-images.githubusercontent.com/27100419/194728342-7f2f1bf6-faa4-445d-a5fc-ac8fc3456a39.png)

**Reply by h3llb3nt** [reply]
okay i will give these steps a try! wonder if this is something aesprite (the program im using to make the gifs) can do as well to eliminate the ezgif work.

**Reply by GrimorJander** [reply]
I have never used the aesprite program.
The only thing I've figured out is how to optimize gifs so they don't show that bug on esp32.
In any case, it is very easy to correct.
I give you an example of a long animation that plays well on a 64x32 screen:
![BadApple](https://user-images.githubusercontent.com/27100419/194728652-9e89b3e8-54f6-483c-bc48-faf0a672772e.gif)

**Reply by AidenWolf001** [reply]
Thank you. This problem has been bothering me for a long time, and finally I have found a solution. Sincere thanks again!
